File: python-refactor/Code/pre_03_station/JP_station_01_raw2mat_soramame.m
clear all;clc; close all;

% for local
% path_data = '//10.72.26.46/irisnas6/Data/In_situ/AirQuality_Japan/';
% path = '//10.72.26.56/irisnas5/Data/Station/Station_JP/';
% addpath(genpath('//10.72.26.56/irisnas5/Data/matlab_func/'))

path_data = '/share/irisnas6/Data/In_situ/AirQuality_Japan/';
path = '/share/irisnas5/Data/Station/Station_JP/';
% addpath(genpath('/share/irisnas5/Data/matlab_func/'))

header = {'doy','yr','mon','day','KST','SO2','CO','OX','NO2','PM10','PM25','scode'};

for yr = 2017
    for mm=1:12
        list_stn = dir([path_data, num2str(yr),'_soramame/',num2str(yr),num2str(mm,'%02i'),'_00/*.csv']);
        list_stn_f = {list_stn.folder}';
        list_stn = {list_stn.name}';
        
        stn_mm=[];
        for k=1:size(list_stn,1)            
            stn_tbl_temp = readtable([list_stn_f{k},'/',list_stn{k}]);
            
            scode = table2array(stn_tbl_temp(:,1));
            dstr = table2array(stn_tbl_temp(:,2));
            dvec = datevec(dstr,'yyyy/mm/dd');
            dnum = datenum(dstr,'yyyy/mm/dd');
            doy_000 = datenum([yr,0,0,0,0,0]);
            doy = dnum-doy_000;
            stn_value = table2array(stn_tbl_temp(:,3:15));
            
            data = [doy,dvec(:,1:3),stn_value(:,[1,2,6,7,4,11,12]),scode];
            % {'doy','yr','mon','day','KST','SO2','CO','OX','NO2','PM10','PM25','scode'};

            stn_mm = [stn_mm; data];
        end
        save([path,'stn_code_data/stn_code_data_',num2str(yr),'_',num2str(mm,'%02i'),'.mat'],'stn_mm','-v7.3');
        disp(mm)
    end        
    
    disp('Stack monthly data to yearly data')
    stn_yr = [];
    for mm=1:12
        load([path,'stn_code_data/stn_code_data_',num2str(yr),'_',num2str(mm,'%02i'),'.mat']);
        stn_yr=[stn_yr;stn_mm];
    end
    save([path, 'stn_code_data/stn_code_data_',num2str(yr),'_2.mat'],'stn_yr','-v7.3');
    disp(yr)
end

File: python-refactor/Code/pre_03_station/JP_station_01_raw2mat_soramame.py
### Package Import
import sys
import os
base_dir = os.environ['GEMS_HOME']
project_path = os.path.join(base_dir, 'python-refactor')
sys.path.insert(0, project_path)
from Code.utils import matlab

import copy
import numpy as np
import pandas as pd
import glob
import scipy.io as sio

### Setting path
data_base_dir = os.path.join('/data2', 'sehyun', 'Data')
path_in_situ = os.path.join(data_base_dir, 'Raw') 
path_station = os.path.join(data_base_dir, 'Preprocessed_raw', 'Station') 
path_stn_jp = os.path.join(path_station, 'Station_JP')

header = ['doy','yr','mon','day','KST','SO2','CO','OX','NO2','PM10','PM25','scode']

YEARS = [2016]
for yr in YEARS:
    for mm in range(1, 12+1):
        file_list = glob.glob(os.path.join(path_in_situ, 'AirQuality_Japan', f'{yr}_soramame', f'{yr}{mm:02d}_00', '*.csv'))
        file_list.sort()
        stn_mm = None
        for fname in file_list:
            stn_tbl_temp = pd.read_csv(fname, encoding='latin1')
            stn_tbl_temp = stn_tbl_temp.values
            scode = stn_tbl_temp[:,0]
            
            dstr = [str(int(x)) for x in stn_tbl_temp[:,1]]
            dstr = [f'{d[:4]}{d[5:7]}{d[8:]}' for d in dstr] # yyyy/mm/dd -> yyyymmdd
            dvec = [(int(d[:4]), int(d[4:6]), int(d[6:])) for d in dstr] 
            dvec = np.array(dvec)
            data_datenum = [matlab.datenum(val) for val in dstr]
            doy_000 = matlab.datenum(f'{yr}00000')
            doy = np.array([val-doy_000 for val in data_datenum])
            
            stn_value = stn_tbl_temp[:,2:15]
            
            data = np.hstack([doy.reshape(-1,1), dvec, stn_value[:,[0,1,5,6,3,10,11]], scode.reshape(-1,1)])
            # {'doy','yr','mon','day','KST','SO2','CO','OX','NO2','PM10','PM25','scode'}
            if stn_mm is None: stn_mm = data
            else: stn_mm = np.vstack([stn_mm, data])
        sio.savemat(os.path.join(path_stn_jp, 'stn_code_data', f'stn_code_data_{yr}_{mm:02d}.mat'),{'stn_mm':stn_mm},
                   do_compression=True)
        print (mm)
    
    print ('Stack monthly data to yearly data')
    stn_yr = None
    for mm in range(1, 12+1):
        stn_mm = matlab.loadmat(os.path.join(path_stn_jp,'stn_code_data', f'stn_code_data_{yr}_{mm:02d}.mat'))['stn_mm']
        if stn_yr is None: stn_yr = stn_mm
        else: stn_yr= np.vstack([stn_yr, stn_mm])
    sio.savemat(os.path.join(path_stn_jp, 'stn_code_data', f'stn_code_data_{yr}_2.mat'),{'stn_yr':stn_yr},
                   do_compression=True)
    print (yr)